File: src_python/dihed/qnmat/qnmat_3.py
import sys
import numpy as np
import cython
from typing import Self
from numpy.typing import NDArray

import crsys as crs
import qnnum as qnn
import qnvec as qnv
import qnndarray as qna
from numpy.typing import(NDArray)

def sub(ma1:Qnmat, ma2:Qnmat) -> Qnmat:
    #a=np.empty(mat1.shape, dtype=qnn.Qnnum)
    shape=ma1.shape
    n1=len(ma1.shape)
    n2=len(ma2.shape)
    #N=ma1[0][0].N
    a=Qnmat(shape)
    if(n1==1 and n2==1): # vectors
        for i in range(shape[0]):
            a[i]=ma1[i]-ma2[i]  #add(v1[i],v2[i])
        return a
    elif(n1==2 and n2==2): # matrices
        for i in range(shape[0]):
            for j in range(shape[1]):
                a[i][j]=ma1[i][j]-ma2[i][j]  #add(v1[i],v2[i])
        return a 
    
def iadd(ma1:Qnmat, ma2:Qnmat) -> Qnmat:
    ma1=add(ma1,ma2)
    return ma1

def isub(ma1:Qnmat, ma2:Qnmat) -> Qnmat:
    ma1=sub(ma1,ma2)
    return ma1

def mul_scl(ma1: Qnmat, scl: qnn.Qnnum) -> Qnmat:
    ndim=len(ma1.shape)
    if ndim==1:
        for i in range(ma1.shape[0]):
            ma1[i]=ma1[i]*scl
    if ndim==2:
        for i in range(ma1.shape[0]):
            for j in range(ma1.shape[1]):
                ma1[i][j]=ma1[i][j]*scl
    return ma1

    
# for ma1@ma2 (ma1 and ma2 should be qnvec or qnmat)
def mul(ma1: Qnmat, ma2: Qnmat) -> Qnmat: 
    ndm1=len(ma1.shape)
    ndm2=len(ma2.shape)
    #print("ndm1=",ndm1,"ndm2=",ndm2)  # for test
    if ndm1==1 and ndm2==1:  # dot product
        if ma1.shape[0]==ma2.shape[0]:
            ma3=qnn.zero()
            for i in range(ma1.shape[0]):
                ma3+=ma1[i]*ma2[i]
        else:
            print("ma1.shape[0] should be equal to ma2.shape[0] for ndim1==ndim2"); exit()
    elif ndm1==1 and ndm2==2:  # vec*matrix
        if ma1.shape[0]==ma2.shape[0]:
            ma3=qnv.zerov(ma2.shape[1])
            for j in range(ma2.shape[1]):
                for i in range(ma1.shape[0]):
                    ma3[j]+=ma1[i]*ma2[i][j]
        else:
            print("ma1.shape[0] should be equal to ma2.shape[0] for ndim1=1 and ndim2=2"); exit()
    elif ndm1==2 and ndm2==1:  # matrix*vec
        if ma1.shape[1]==ma2.shape[0]:
            ma3=qnv.zerov(ma1.shape[0])
            for j in range(ma2.shape[0]):
                for i in range(ma1.shape[0]):
                    ma3[j]+=ma1[j][i]*ma2[i]
        else:
            print("ma1.shape[1] should be equal to ma2.shape[0] for ndim1=2 and ndim2=1"); exit()
    elif ndm1==2 and ndm2==2:
        if ma1.shape[0]==ma1.shape[1] and ma2.shape[0]==ma2.shape[1] and ma1.shape[0]==ma2.shape[0]:
            ma3=qnv.zerov(ma1.shape)
            for k in range(ma2.shape[0]):
                for j in range(ma1.shape[0]):
                    for i in range(ma1.shape[0]):
                        ma3[k][j]+=ma1[k][i]*ma2[i][j]
        else:
             print("square matrix is assumed for ndm1=ndm2=2"); exit()

    #ma3=np.matmul(ma1,ma2,dtype=qnn.Qnnum)
    return ma3

# for similarity transformation
# not confirmed yet
def pow(ma: Qnmat, n_: int) -> Qnmat:
    """
    """
    (mx,my)=ma.shape
    #N=ma[0][0].N
    if mx==my:
        if n_==0:
            return np.identity(mx)
#        elif n<0:
#            tmp=unitm(n,N)
#            mai = copy(ma) # copy for qnmatinv
#            qmt.qnmatinv(mai,n) #???
#            for i in range(-n):
#                tmp=np.dot(tmp,inva)
#            return tmp
        else:
            tmp=np.unitm(n_,N)
            for i in range(n_):
                tmp=np.dot(tmp,ma)
            return tmp
    else:
        print('matrix has not regular shape')
        exit()

File: src_python/dihed/qnmat/cpp/qnmat_3.cpp
#include <iostream>
#include <vector>
#include <stdexcept>
#include "qnn.h" // Assuming these headers exist for Qnnum, Qnmat, Qnvec
#include "qnv.h"
#include "qna.h"

using namespace std;

Qnmat sub(const Qnmat& ma1, const Qnmat& ma2) {
    auto shape = ma1.shape();
    int n1 = ma1.ndim();
    int n2 = ma2.ndim();
    Qnmat a(shape);
    if (n1 == 1 && n2 == 1) { // vectors
        for (int i = 0; i < shape[0]; i++) {
            a[i] = ma1[i] - ma2[i];
        }
        return a;
    } else if (n1 == 2 && n2 == 2) { // matrices
        for (int i = 0; i < shape[0]; i++) {
            for (int j = 0; j < shape[1]; j++) {
                a[i][j] = ma1[i][j] - ma2[i][j];
            }
        }
        return a;
    }
    throw invalid_argument("Invalid dimensions for subtraction");
}

Qnmat iadd(Qnmat& ma1, const Qnmat& ma2) {
    ma1 = add(ma1, ma2);
    return ma1;
}

Qnmat isub(Qnmat& ma1, const Qnmat& ma2) {
    ma1 = sub(ma1, ma2);
    return ma1;
}

Qnmat mul_scl(Qnmat& ma1, const Qnnum& scl) {
    int ndim = ma1.ndim();
    if (ndim == 1) {
        for (int i = 0; i < ma1.shape()[0]; i++) {
            ma1[i] = ma1[i] * scl;
        }
    } else if (ndim == 2) {
        for (int i = 0; i < ma1.shape()[0]; i++) {
            for (int j = 0; j < ma1.shape()[1]; j++) {
                ma1[i][j] = ma1[i][j] * scl;
            }
        }
    }
    return ma1;
}

Qnmat mul(const Qnmat& ma1, const Qnmat& ma2) {
    int ndm1 = ma1.ndim();
    int ndm2 = ma2.ndim();
    Qnmat ma3;

    if (ndm1 == 1 && ndm2 == 1) { // dot product
        if (ma1.shape()[0] == ma2.shape()[0]) {
            ma3 = qnn::zero();
            for (int i = 0; i < ma1.shape()[0]; i++) {
                ma3 += ma1[i] * ma2[i];
            }
        } else {
            throw invalid_argument("ma1.shape[0] should be equal to ma2.shape[0] for ndim1==ndim2");
        }
    } else if (ndm1 == 1 && ndm2 == 2) { // vec*matrix
        if (ma1.shape()[0] == ma2.shape()[0]) {
            ma3 = qnv::zerov(ma2.shape()[1]);
            for (int j = 0; j < ma2.shape()[1]; j++) {
                for (int i = 0; i < ma1.shape()[0]; i++) {
                    ma3[j] += ma1[i] * ma2[i][j];
                }
            }
        } else {
            throw invalid_argument("ma1.shape[0] should be equal to ma2.shape[0] for ndim1=1 and ndim2=2");
        }
    } else if (ndm1 == 2 && ndm2 == 1) { // matrix*vec
        if (ma1.shape()[1] == ma2.shape()[0]) {
            ma3 = qnv::zerov(ma1.shape()[0]);
            for (int j = 0; j < ma2.shape()[0]; j++) {
                for (int i = 0; i < ma1.shape()[0]; i++) {
                    ma3[j] += ma1[j][i] * ma2[i];
                }
            }
        } else {
            throw invalid_argument("ma1.shape[1] should be equal to ma2.shape[0] for ndim1=2 and ndim2=1");
        }
    } else if (ndm1 == 2 && ndm2 == 2) {
        if (ma1.shape()[0] == ma1.shape()[1] && ma2.shape()[0] == ma2.shape()[1] && ma1.shape()[0] == ma2.shape()[0]) {
            ma3 = qnv::zerov(ma1.shape());
            for (int k = 0; k < ma2.shape()[0]; k++) {
                for (int j = 0; j < ma1.shape()[0]; j++) {
                    for (int i = 0; i < ma1.shape()[0]; i++) {
                        ma3[k][j] += ma1[k][i] * ma2[i][j];
                    }
                }
            }
        } else {
            throw invalid_argument("square matrix is assumed for ndm1=ndm2=2");
        }
    }

    return ma3;
}

Qnmat pow(const Qnmat& ma, int n_) {
    auto shape = ma.shape();
    int mx = shape[0];
    int my = shape[1];
    if (mx == my) {
        if (n_ == 0) {
            return Qnmat::identity(mx);
        } else {
            Qnmat tmp = Qnmat::unitm(n_, N); // Assuming unitm is a static method
            for (int i = 0; i < n_; i++) {
                tmp = mul(tmp, ma);
            }
            return tmp;
        }
    } else {
        throw invalid_argument("matrix has not regular shape");
    }
}

